drivers/hv/mshv_vtl_sidecar.c

```diff
@@ -0,0 +1,553 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2024, Microsoft Corporation.
+ *
+ */
+#include <linux/platform_device.h>
+#include <linux/miscdevice.h>
+#include <linux/poll.h>
+#include <linux/mod_devicetable.h>
+#include <linux/mm.h>
+#include <uapi/linux/mshv.h>
+#include <asm/apic.h>
+#include <asm/irq_vectors.h>
+#include <asm/irq.h>
+
+#include "mshv.h"
+
+struct sidecar_control {
+	u32 index;
+	u32 base_cpu;
+	u32 cpu_count;
+	u32 request_vector;
+	u32 response_cpu;
+	u32 response_vector;
+	u32 needs_attention;
+	u8 reserved[36];
+	u8 cpu_status[4032];
+};
+
+#define CPU_STATUS_REMOVED 0
+#define CPU_STATUS_IDLE 1
+#define CPU_STATUS_RUN 2
+#define CPU_STATUS_STOP 3
+#define CPU_STATUS_REMOVE 4
+
+static_assert(sizeof(struct sidecar_control) == 4096);
+
+static LIST_HEAD(sidecar_dev_list);
+
+struct sidecar_dev {
+	struct device *dev;
+	struct sidecar_control *control;
+	struct wait_queue_head wait;
+	struct resource *shmem_pages;
+	u32 base_cpu;
+	u32 cpu_count;
+	u32 index;
+	u32 per_cpu_shmem;
+	u8 *vp_state;
+	u32 needs_vp_scan;
+	u32 num_vps_stopped;
```

Comment:
this can be a `atomic_t num_vps_stopped`, and then you can use `atomic_add_return()` to replace `xadd()`.